File: generate_chord_table.py
#!/usr/bin/env python3
# generate_chord_table.py
# Generates a CSV of all chords and fingerings for the Venezuelan Cuatro.
# Usage: python3 generate_chord_table.py > chords.csv

import json, csv, sys
from itertools import product

with open('cuatro_acordes.json', encoding='utf-8') as f:
    CFG = json.load(f)

STRING_BASES = CFG['algoritmo']['bases_cuerdas']  # [11, 6, 2, 9] → B, F#, D, A
W = CFG['algoritmo']['pesos']

def nota_a_num(nota):
    s = CFG['notas']['sostenidos']
    b = CFG['notas']['bemoles']
    return s.index(nota) if nota in s else b.index(nota)

def nombre_acorde(nota, calidad):
    q = CFG['calidades'].get(calidad, {})
    return nota + q.get('simbolo', calidad)

def calcular_digitacion(raiz_num, calidad):
    q = CFG['calidades'].get(calidad)
    if not q:
        return [0, 0, 0, 0]
    intervalos = q['intervalos']
    notas_acorde = set((raiz_num + i) % 12 for i in intervalos)

    # Check overrides
    nota_s = CFG['notas']['sostenidos'][raiz_num]
    key_s = nombre_acorde(nota_s, calidad)
    if key_s in CFG['digitaciones_override']:
        return CFG['digitaciones_override'][key_s]
    nota_b = CFG['notas']['bemoles'][raiz_num]
    key_b = nombre_acorde(nota_b, calidad)
    if key_b in CFG['digitaciones_override']:
        return CFG['digitaciones_override'][key_b]

    # Build fret options per string
    opciones = []
    for base in STRING_BASES:
        opts = [f for f in range(10) if (base + f) % 12 in notas_acorde]
        opciones.append(opts if opts else [0])

    mejor = None
    mejor_punt = float('inf')

    for combo in product(*opciones):
        cubiertos = set((STRING_BASES[i] + combo[i]) % 12 for i in range(4))
        if raiz_num % 12 not in cubiertos:
            continue

        faltantes  = len([n for n in notas_acorde if n not in cubiertos])
        presionados = [f for f in combo if f > 0]
        distinto   = len(set(presionados))
        fspan      = (max(presionados) - min(presionados)) if presionados else 0
        fp = [i for i, f in enumerate(combo) if f > 0]
        op = [i for i, f in enumerate(combo) if f == 0]
        guia     = len([i for i in op if not fp or i < fp[0]])
        arrastre = len([i for i in op if not fp or i > fp[-1]])
        medio    = len([i for i in op if fp and fp[0] < i < fp[-1]])
        min_traste  = min(presionados) if presionados else 0
        barre_puro  = presionados[0] if presionados and len(set(presionados)) == 1 else 0
        mx = max(combo)
        sm = sum(combo)

        punt = (faltantes  * W['nota_faltante']
              + distinto   * W['distintas']
              + guia * min_traste * W['escala_guia']
              + arrastre   * W['arrastre']
              + medio      * W['medio']
              + barre_puro * W['barre_puro']
              + fspan      * W['amplitud']
              + mx         * W['maximo']
              + sm)

        if punt < mejor_punt:
            mejor_punt = punt
            mejor = list(combo)

    return mejor or [0, 0, 0, 0]


writer = csv.writer(sys.stdout)
writer.writerow(['Chord', 'Root (EN)', 'Root (ES)', 'Quality', 'Symbol', 'B', 'F#', 'D', 'A', 'Fingering'])

for nota in CFG['notas']['sostenidos']:
    raiz_num = nota_a_num(nota)
    nota_es  = CFG['notas']['nombres_display'].get(nota, nota)

    for cal_key, cal in CFG['calidades'].items():
        symbol = cal['simbolo'] or cal_key
        chord  = nota + symbol
        dig    = calcular_digitacion(raiz_num, cal_key)
        writer.writerow([chord, nota, nota_es, cal['nombre'], symbol,
                         dig[0], dig[1], dig[2], dig[3], ''.join(map(str, dig))])

File: generate_chord_table.js
// generate_chord_table.js
// Generates a CSV of all chords and fingerings for the Venezuelan Cuatro.
// Usage: node generate_chord_table.js > chords.csv

const fs = require('fs');
const CFG = JSON.parse(fs.readFileSync('./cuatro_acordes.json', 'utf8'));

const STRING_BASES = CFG.algoritmo.bases_cuerdas; // [11, 6, 2, 9] → B, F#, D, A

function notaANum(nota) {
  const idx = CFG.notas.sostenidos.indexOf(nota);
  if (idx >= 0) return idx;
  return CFG.notas.bemoles.indexOf(nota);
}

function nombreAcorde(nota, calidad) {
  const q = CFG.calidades[calidad];
  if (!q) return nota;
  return nota + q.simbolo;
}

function calcularDigitacion(raizNum, calidad) {
  const q = CFG.calidades[calidad];
  if (!q) return [0, 0, 0, 0];
  const intervalos = q.intervalos;
  const notasAcorde = new Set(intervalos.map(i => ((raizNum + i) % 12 + 12) % 12));

  // Check overrides first (sharp name, then flat name)
  const notaS = CFG.notas.sostenidos[raizNum];
  const overrideS = CFG.digitaciones_override[nombreAcorde(notaS, calidad)];
  if (overrideS) return overrideS;
  const notaB = CFG.notas.bemoles[raizNum];
  const overrideB = CFG.digitaciones_override[nombreAcorde(notaB, calidad)];
  if (overrideB) return overrideB;

  const w = CFG.algoritmo.pesos;

  const opciones = STRING_BASES.map(base =>
    Array.from({ length: 10 }, (_, f) => f).filter(f => notasAcorde.has((base + f) % 12))
  );
  opciones.forEach(op => { if (op.length === 0) op.push(0); });

  let mejor = null, mejorPuntuacion = Infinity;

  function iterar(s, combo) {
    if (s === 4) {
      const cubiertos = new Set(STRING_BASES.map((b, i) => (b + combo[i]) % 12));
      if (!cubiertos.has(raizNum % 12)) return;
      const faltantes = [...notasAcorde].filter(n => !cubiertos.has(n)).length;
      const presionados = combo.filter(f => f > 0);
      const distinto = new Set(presionados).size;
      const fspan = presionados.length > 0 ? Math.max(...presionados) - Math.min(...presionados) : 0;
      const fp = combo.map((f, i) => f > 0 ? i : -1).filter(i => i >= 0);
      const op = combo.map((f, i) => f === 0 ? i : -1).filter(i => i >= 0);
      const guia     = op.filter(i => fp.length === 0 || i < fp[0]).length;
      const arrastre = op.filter(i => fp.length === 0 || i > fp[fp.length - 1]).length;
      const medio    = op.filter(i => fp.length > 0 && i > fp[0] && i < fp[fp.length - 1]).length;
      const minTraste  = presionados.length > 0 ? Math.min(...presionados) : 0;
      const barrePuro  = (presionados.length > 0 && new Set(presionados).size === 1) ? presionados[0] : 0;
      const mx = Math.max(...combo);
      const sm = combo.reduce((a, b) => a + b, 0);
      const puntuacion = faltantes * w.nota_faltante
        + distinto * w.distintas
        + guia * minTraste * w.escala_guia
        + arrastre * w.arrastre
        + medio * w.medio
        + barrePuro * w.barre_puro
        + fspan * w.amplitud
        + mx * w.maximo
        + sm;
      if (puntuacion < mejorPuntuacion) {
        mejorPuntuacion = puntuacion;
        mejor = [...combo];
      }
      return;
    }
    for (const f of opciones[s]) {
      combo[s] = f;
      iterar(s + 1, combo);
    }
  }
  iterar(0, [0, 0, 0, 0]);
  return mejor || [0, 0, 0, 0];
}

// CSV header
// Strings are ordered B · F# · D · A (high to low)
const rows = ['Chord,Root (EN),Root (ES),Quality,Symbol,B,F#,D,A,Fingering'];

for (const nota of CFG.notas.sostenidos) {
  const raizNum = notaANum(nota);
  const notaES  = CFG.notas.nombres_display[nota] || nota;

  for (const [calKey, cal] of Object.entries(CFG.calidades)) {
    const symbol = cal.simbolo || calKey;
    const chord  = nota + symbol;
    const dig    = calcularDigitacion(raizNum, calKey);
    rows.push(`${chord},${nota},${notaES},${cal.nombre},${symbol},${dig[0]},${dig[1]},${dig[2]},${dig[3]},${dig.join('')}`);
  }
}

process.stdout.write(rows.join('\n') + '\n');
